pkg/vsphere/rbac/rbac.go

```diff
@@ -163,6 +169,53 @@ func (am *AuthzManager) PrincipalBelongsToGroup(ctx context.Context, group strin
 	return false, nil
 }
 
+// PrincipalHasReadOnlyPrivs returns true if the user (principal) in the AuthzManager has
+// at least a read-only role on the input ref object, false otherwise.
+func (am *AuthzManager) PrincipalHasReadOnlyPrivs(ctx context.Context, ref types.ManagedObjectReference) (bool, error) {
+	perms, err := am.GetPermissions(ctx, ref)
+	if err != nil {
+		return false, err
+	}
+
+	roles, err := am.RoleList(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	opsUser := strings.ToLower(am.Principal)
+
+	// These vars keep track of whether the corresponding privileges have been found.
+	var anon, read, view bool
+
+	for _, perm := range perms {
```

Comment:
how large this array can be?

Reply:
A permission entry contains the user and role. We're querying all entries for the datacenter ref here, so in a prod env with multiple users and roles, this array can have a decent amount of entries. Can't put a number on how many elements there can be, but it's handy to check the flags at the bottom of each iteration to avoid making extraneous calls to `formatPrincipal` and `roles.ById` if we've already met the criteria.